Tzkt.Api/Websocket/Processors/CyclesProcessor.cs

```diff
@@ -0,0 +1,150 @@
+﻿using System;
+using System.Linq;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.SignalR;
+using Microsoft.Extensions.Logging;
+
+using Tzkt.Api.Repositories;
+using Tzkt.Api.Services.Cache;
+
+namespace Tzkt.Api.Websocket.Processors
+{
+    public class CycleProcessor<T> : IHubProcessor where T : Hub
+    {
+        #region static
+        const string CycleGroup = "cycle";
+        const string CycleChannel = "cycle";
+        static readonly SemaphoreSlim Sema = new(1, 1);
+        static readonly Dictionary<int, HashSet<string>> DelaySubs = new();
+        #endregion
+
+        readonly StateCache StateCache;
+        readonly StateRepository StateRepo;
+        readonly IHubContext<T> Context;
+        readonly ILogger Logger;
+
+        private int cycleStartLevel = 0;
+        private int lastProcessedCycle = 0;
+        
+        public CycleProcessor(StateCache cache, StateRepository repo, IHubContext<T> hubContext, ILogger<CycleProcessor<T>> logger)
+        {
+            StateCache = cache;
+            StateRepo = repo;
+            Context = hubContext;
+            Logger = logger;
+        }
+
+        public async Task OnStateChanged()
+        {
+            var sendings = new List<Task>(2);
+            try
+            {
+                await Sema.WaitAsync();
+              
+                if (lastProcessedCycle != StateCache.Current.Cycle)
+                    cycleStartLevel = StateCache.Current.Level;
```

Comment:
So, if I start/restart the API, the `cycleStartLevel` will contain a wrong value (the current level), and the API will send false notifications.

Reply:
I switched to cycle data from CyclesRepository. It should be better now